### DISCUSSION TITLE
About the auto accumulation reduction in AD

### DISCUSSION THREAD

**Discussion by lesphere** [mitsuba-renderer/mitsuba3#995] [Q&A] [answered]
Hi, 

I cannot find which part of the code do the accumulation as described below. Could you please give me some advice about where it is?

> However, if at any point in the AD graph all the threads traverse some shared vertex of width 1, then the result you'll be computing is most likely not what you'd want because the intermediary result of that shared vertex would again be the accumulation of all threads.

**Comment by njroussel** [comment, marked as answer]
Hi @lesphere 

This happens deep within Dr.Jit's AD system. This is the relevant piece of code: https://github.com/mitsuba-renderer/drjit/blob/master/src/autodiff/autodiff.cpp#L237-L350. As you'll see there are two sections that handle the reduction when the source has size N and the sink has size 1.
